scripts/sub-digest/subdigest.py

```diff
@@ -130,15 +131,17 @@ def _get_nonselection(self):
                 if i not in self.selection]
 
     def _process_selection(self, f):
+        lines = self._get_selection()
+
         if self.selection is None:
-            f(self._get_section())
+            indices = range(len(lines))
         else:
             indices = sorted(self.selection)
-            lines = self._get_selection()
-            f(lines)
-            section = self._get_section()
-            for i, line in zip(indices, lines):
-                section[i] = line
+
+        f(lines)
```

Comment:
Reading this code properly, I'm convinced that `_process_selection` should call the provided callback with each line as the argument and not a list of lines so that it would be sortable since that would technically also allow modifying the list's size (by adding or removing entries) and the following code wouldn't handle that properly.

Sorting should then use a `_sort_selection` handler that accepts a `key` function as the parameter which will then be used on the list that is extracted similarly to how it happens in `_process_selection` currently because that is a still a decent way to implement this.

Now that's all about internal code, but I like clean APIs also for internal calls. It would also have likely made the bug being fixed here more obvious.

Reply:
As a clarification, that's not meant as a request to change the PR. I was just rambling about it as I was reading the code. You're still free to implement it that way, of course, and split the two different kinds of usages of `_process_lines`.

Always returning a list in `_get_selection` may not be a bad thing either but it does have a tiny bit of overhead when it may not be needed most of the time.